### DISCUSSION TITLE
Custom api endpoint, Proxy, returning websocket instance..

### DISCUSSION THREAD

**Discussion by cpt-n3mo** [directus/directus#8746] [Archive]
Good day..

I have been working on a custom module and endpoint for a static site generator, in combination with docker-compose and this works like a charm, also this module includes a iframe to a nuxt dev instance running on the same docker network,
im now routing the dev server though nginx via a subdomain, this is somewhat annoying and i was wondering if there is a way to setup a local proxy so that i can route http://dev_server:3000 from the docker network to some endpoint so that it can access this inside the client side module. 

On the static generator part: 
have set up a custom endpoint to call the script on the server side, now the down side is that i have to wait until the script is finished until i can sent back the stdout of the script.
im thinking a websocket would be nice to get data back directly, i could setup a the websocket on the server side and then route it though a subdomain in nginx, but the big down side to this is that it is accessible from the outside and adds more subdomains.
is there a way to send a websocket back over the custom endpoint.? so that i dont have to route it through nginx..?
or is there a way to setup a internal proxy so i can redirect it to the local websocket,  basically the same a iframe idea. 

kind regards.

**Comment by rijkvanzanten** [comment]
G'day!

I'm having a bit of trouble trying to understand exactly what the setup is.. 

> have set up a custom endpoint to call the script on the server side, now the down side is that i have to wait until the script is finished until i can sent back the stdout of the script.

Just for my own understanding, what does this endpoint do?

> is there a way to send a websocket back over the custom endpoint.?

I'm not sure if this is as straightforward as it sounds. A websocket connection is done on a websocket server, which isn't the same as the http server. This would mean you'd have to have a listener on a different port next to regular Directus. Alternatively, you could try to tap into Directus' express server in the custom endpoint, and try to attach the upgrade listener of the server in order to serve both the http(s) and ws(s) servers on the same port, but that still doesn't necessarily change anything in how the traffic is then routed from Nginx to Directus / your web socket server.

**Reply by cpt-n3mo** [reply]
Good day..

thank you for your quick response  :)
And Sorry for being unclear on the setup..

endpoint `/generate/gen` spawns a process on the server, that generates a static site from a nuxt app inside a docker container and then serves it on nginx.
after the process exits i send back the stdout or stderr depending on exit code from the spawned process.

as generation can take some time, it would be nice to get the data back in real time. so this is why i was thinking about a websocket. if there is a better way of getting the same functionality then that would be fine as well. i just though a websocket would be easy to setup if it could be proxied ofc :)

> This would mean you'd have to have a listener on a different port next to regular Directus. Alternatively, you could try to tap into Directus' express server in the custom endpoint, and try to attach the upgrade listener of the server in order to serve both the http(s) and ws(s) servers on the same port

indeed i have setup a websocket using socket.io on port 8000. i was just reading about `http-proxy` and that it can proxy websockets as well, so i might take a look at that.. 

i just managed to proxy the dev server inside the docker network. so i can call `src="/generate/iframe"` on the iframe client side using `express-http-proxy` 
`  router.use('/iframe', proxy('localAdress:port')) `

 so i might be able to do something similar for the websocket.
 kind regards.

**Reply by rijkvanzanten** [reply]
Interesting! So you're effectively statically generating the website in runtime instead of build time. Haven't heard of people doing that before, lemme know how it goes!

**Reply by joselcvarela** [reply]
Hello @cpt-n3mo 
Alternatively to websockets, you can use server-side events. This is a one way communication (from server to client) and is supported by browser via EventSource https://developer.mozilla.org/en-US/docs/Web/API/EventSource.
Though you will need to use a package or implement it by yourself on server side.

Also you can make the long request via fetch and proceed with other requests. Remember, if you use async/await for fetch, this long fetch shouldn't block the others, i.e. you cannot await this fetch on same block of other requests.

**Reply by cpt-n3mo** [reply]
Good Day!

> Interesting! So you're effectively statically generating the website in runtime instead of build time. Haven't heard of people doing that before, lemme know how it goes!

indeed! im effectively mounting de nuxt app folder inside the directus container so i have access to it from within directus. at the same time this nuxt app is running in dev mode and iframed,  components are marked with a attr with its component id. so i can inject some edit buttons inside the iframe, the idea is to make a click and edit from the dev page, and then when one is happy with the dev version on can push the changes to the static site,  this gives no down time and makes editing static sites super easy.

>Hello @cpt-n3mo
Alternatively to websockets, you can use server-side events. This is a one way communication (from server to client) and is supported by browser via EventSource https://developer.mozilla.org/en-US/docs/Web/API/EventSource.
Though you will need to use a package or implement it by yourself on server side.
Also you can make the long request via fetch and proceed with other requests. Remember, if you use async/await for fetch, this long fetch shouldn't block the others, i.e. you cannot await this fetch on same block of other requests.

OH! This is awesome! did not know about this. this is even better as this is unidirectional, and deff easier to implement then the websocket! thx!! this will help allot!!

**Reply by martinemmert** [reply]
@cpt-n3mo have you had success using SSE on this one? I am currently trying to get them running but having a hard time getting the simplest example to run :-|
